src/Lucene.Net/Search/FieldCacheImpl.cs

```diff
@@ -73,11 +74,17 @@ private void  Init()
             _stringIndexCache = new StringIndexCache(this);
         }
 
-        public virtual void  PurgeAllCaches()
+        public virtual IDisposable PurgeAllCaches()
         {
             lock (this)
             {
+                // PurgeAllCaches is replacing the cache with a new one (without actually releasing any memory).
+                // When the GC will run, the finalizer of the Segments will be executed and release the unmanaged memory.
+                // We'll return the old StringIndexCache and let the caller decide if he wants to dispose it sooner
+
+                var copy = _stringIndexCache;
                 Init();
```

Comment:
Pretty sure that we aren't locking on `this` on each read, so note that the `lock` isn't actually meaningful.
You may be getting partial reads here.


In this scenario, I don't think it matters, but something to be aware of

Reply:
I want to lock at the RavenDB level, not here.
When I'm not running any queries.

Reply:
who else is doing the locking?

Reply:
This was the original code in Lucene. There is no need to do the locking if we are the only caller.

The idea is that the `Init` will be called when we aren't running any query.
The lock will be on the RavenDB side.

https://github.com/ravendb/ravendb/pull/16646/files#diff-3760b1669cd9af0aab370cfc92d98e3898fed9cea969a44cb19e7430aa953509R37

https://github.com/ravendb/ravendb/pull/16646/files#diff-3760b1669cd9af0aab370cfc92d98e3898fed9cea969a44cb19e7430aa953509R46

Reply:
LGTM, then